File: Design War Measurement System/Code/Java/src/fayelab/ddd/measure/vector/second/VecOp.java
package fayelab.ddd.measure.vector.second;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.stream.Collector;
import java.util.stream.Collectors;

public class VecOp
{
    public static boolean equal(List<Integer> firstVec, List<Integer> secondVec)
    {
        validateNotNull(firstVec, secondVec);
        
        if(firstVec.size() != secondVec.size())
        {
            return false;
        }
        
        return operate(firstVec, secondVec, (x, y) -> x == y, Collectors.reducing(true, (b1, b2) -> b1 && b2));
    }
    
    public static List<Integer> add(List<Integer> augendVec, List<Integer> addendVec)
    {
        validate(augendVec, addendVec);
        
        return operate(augendVec, addendVec, (x, y) -> x + y,  Collectors.toList());
    }
    
    public static int dotProduct(List<Integer> rowVec, List<Integer> colVec)
    {
        validate(rowVec, colVec);
        
        return operate(rowVec, colVec, (x, y) -> x * y, Collectors.reducing(0, Integer::sum));
    }
    
    public static List<Integer> div(List<Integer> dividendVec, List<Integer> divisorVec)
    {
        validate(dividendVec, divisorVec);
        validateDivisorVec(divisorVec);
        
        List<Integer> result = zeroList(dividendVec.size());
        int carry = 0;
        
        for(int i = dividendVec.size() - 1; i > 0; i--)
        {
            int dividend = dividendVec.get(i) + carry;
            int divisor = divisorVec.get(i);
            int remainder = dividend % divisor;
            result.set(i, remainder);
            carry = dividend / divisor;
        }
        result.set(0, carry + dividendVec.get(0));

        return result;
    }
        
    public static List<Integer> scale(int c, List<Integer> vec)
    {
        validateNotNull(vec);
        
        return vec.stream().map(x -> c * x).collect(Collectors.toList());
    }

    static void validate(List<Integer> firstVec, List<Integer> secondVec)
    {
        validateNotNull(firstVec, secondVec);
        validateSameSize(firstVec, secondVec);
    }

    private static void validateNotNull(List<Integer> firstVec, List<Integer> secondVec)
    {
        if(null == firstVec)
        {
            throw new IllegalParametersException("The first vector is null.");
        }
        
        if(null == secondVec)
        {
            throw new IllegalParametersException("The second vector is null.");
        }
    }
    
    private static void validateNotNull(List<Integer> vec)
    {
        if(null == vec)
        {
            throw new IllegalParametersException("The vector is null.");
        }
    }

    private static void validateSameSize(List<Integer> firstVec, List<Integer> secondVec)
    {
        if(firstVec.size() != secondVec.size())
        {
            throw new IllegalParametersException("The sizes of two vectors are different.");
        }
    }

    private static void validateDivisorVec(List<Integer> divisorVec)
    {
        if(divisorVec.isEmpty())
        {
            throw new IllegalParametersException("The divisor vector is empty.");
        }
        
        if(divisorVec.contains(0))
        {
            throw new IllegalParametersException("The divisor vector which includes 0 components is illegal.");
        }
    }
    
    private static <R1, A, R2> R2 operate(List<Integer> firstVec, List<Integer> secondVec, 
            BiFunction<Integer, Integer, R1> biFunction, Collector<? super R1, A, R2> collector)
    {
        List<R1> list = new ArrayList<>();
        
        for(int i = 0; i < firstVec.size(); i++)
        {
            list.add(biFunction.apply(firstVec.get(i), secondVec.get(i)));
        }
        
        return list.stream().collect(collector);
    }
    
    private static List<Integer> zeroList(int size)
    {
        List<Integer> result = new ArrayList<>();
        for(int i = 0; i < size; i++)
        {
            result.add(0);
        }
        return result;
    }
}

File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/arithmetic/vecop.py
def equal(v1, v2):
    return v1 == v2

def add(v1, v2):
    if len(v1) == len(v2):
        return [vc1 + vc2 for (vc1, vc2) in zip(v1, v2)]

    return None

def dotproduct(v1, v2):
    if len(v1) == len(v2):
        return sum([vc1 * vc2 for (vc1, vc2) in zip(v1, v2)])

    return None

def div(v1, v2):
    if len(v1) == len(v2) and len(v2) != 0 and (0 not in v2):
        return _div(v1, v2)

    return None

def _div(v1, v2):
    result = []
    carry = 0
    for (vc1, vc2) in zip(v1[::-1], v2[::-1]):
        dividend = vc1 + carry
        divisor = vc2
        remainder = dividend % divisor
        result.append(remainder)
        carry = dividend / divisor
    return result[::-1]

def _div2(v1, v2):
    return reduce(_acc_div2, zip(v1[::-1], v2[::-1]), ([], [0]))[0][::-1]

def _acc_div2((v, wcarry), (vc1, vc2)):
    dividend = vc1 + wcarry[0]
    divisor = vc2
    remainder = dividend % divisor
    v.append(remainder)
    wcarry[0] = dividend / divisor
    return v, wcarry

def dotdiv(c, v):
    if len(v) != 0 and (0 not in v):
        return _dotdiv(c, v)

    return None

def _dotdiv(c, v):
    result = []
    remainder = c
    for vc in v:
        quotient = remainder / vc
        remainder = remainder % vc
        result.append(quotient)
    return result

def _dotdiv2(c, v):
    return reduce(_acc_dotdiv2, v, [[], [c]])[0]

def _acc_dotdiv2((v, wc), vc):
    quotient = wc[0] / vc
    remainder = wc[0] % vc
    v.append(quotient)
    wc[0] = remainder
    return v, wc

def scale(c, v):
    return [c * vc for vc in v]